### ISSUE TITLE
"long" math performance

### ISSUE THREAD

**Issue by shannah** [konsoletyper/teavm#86]
I have been doing some performance profiling in Chrome, and I identified one possible bottleneck.  It seems that math on "longs" is pretty slow.  I don't have a solution for this yet - and it's not a critical issue.  One thought is that for most math that occurs, the actual inputs are small enough that they could be done with regular integer math.  If there is an efficient way to check first, this may be some low-hanging fruit to improve performance.

**Comment by shannah** [comment]
This one change leads to a noticeable improvement in the performance of the "spinner" widget.

~~~
diff --git a/teavm-core/src/main/resources/org/teavm/javascript/runtime.js b/tea
index 02dbe2d..d921c30 100644
--- a/teavm-core/src/main/resources/org/teavm/javascript/runtime.js
+++ b/teavm-core/src/main/resources/org/teavm/javascript/runtime.js
@@ -570,6 +570,9 @@ function Long(lo, hi) {
     this.hi = hi | 0;
 }
 Long.prototype.toString = function() {
+    if (this.hi===0){
+        return ""+Long_toNumber(this);
+    }
     var result = [];
     var n = this;
     var positive = Long_isPositive(n);
~~~

But this is just superficial.  I think there is probably a some more comprehensive changes that would have an even more dramatic effect if we hit the LongInt_div function (and possibly a few others) directly.  I tinkered with the LongInt_div function but I don't fully understand what the outputs have to be.   I understand the quotient part but it seems that the the first parameter is also mutated by the function.

**Comment by konsoletyper** [comment]
I tried to fall-back into native operations for smaller values: https://github.com/konsoletyper/teavm/commit/0572123c0e31d4883593392a3ece4c0c74918922
This didn't increase performance of spinnner, since it continues to divide large numbers to compute dates from timestamps. LongInt_div uses Knuth's division algoritm. Remember, you've learned at school how to divide large number in decimal basis. This algoritms does the same using 2^16 basis.

**Comment by shannah** [comment]
Sorry for the red herring.  I ran the Chrome profiler (which btw works great now with the new "better-async" branch) during scrolling of the spinner to try to see where the bottleneck was.  I noticed that a lot of time was spent in LongInt_div.  When I originally changed the toString() method as I posted, and tried, the spinner seemed to be working better.    But when I reverted back it was still working pretty well, so it could have been inconsistencies due to CPU load at the time, or JIT, etc..
